gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java

```diff
@@ -1494,65 +1498,72 @@ private Credential getCredential(Configuration config)
   }
 
   /**
    * Generate a {@link Credential} from the internal access token provider based on the service
    * account to impersonate.
    */
   private static Optional<Credential> getImpersonatedCredential(
-      Configuration config, Credential credential) throws IOException {
-    GoogleCloudStorageOptions options =
-        GoogleHadoopFileSystemConfiguration.getGcsFsOptionsBuilder(config)
-            .build()
-            .getCloudStorageOptions();
-
-    String serviceAccountToImpersonate =
-        IMPERSONATION_SERVICE_ACCOUNT_SUFFIX
-            .withPrefixes(CONFIG_KEY_PREFIXES)
-            .get(config, config::get);
-    if (isNullOrEmpty(serviceAccountToImpersonate)) {
-      Optional<String> serviceAccountToImpersonateFromUser =
-          getMatchedServiceAccountToImpersonateFromGroup(
-              options.getUserImpersonationServiceAccounts(),
-              ImmutableList.of(UserGroupInformation.getCurrentUser().getShortUserName()));
-      Optional<String> serviceAccountToImpersonateFromGroup =
-          getMatchedServiceAccountToImpersonateFromGroup(
-              options.getGroupImpersonationServiceAccounts(),
-              ImmutableList.copyOf(UserGroupInformation.getCurrentUser().getGroupNames()));
-      serviceAccountToImpersonate =
-          serviceAccountToImpersonateFromUser.orElse(
-              serviceAccountToImpersonateFromGroup.orElse(null));
+      Configuration config, GoogleCloudStorageFileSystemOptions gcsFsOptions, Credential credential)
+      throws IOException {
+    GoogleCloudStorageOptions options = gcsFsOptions.getCloudStorageOptions();
+    UserGroupInformation currentUser = UserGroupInformation.getCurrentUser();
+    Optional<String> serviceAccountToImpersonate =
+        Stream.of(
+                () ->
+                    getServiceAccountToImpersonateForUserGroup(
+                        USER_IMPERSONATION_SERVICE_ACCOUNT_SUFFIX
+                            .withPrefixes(CONFIG_KEY_PREFIXES)
+                            .getPropsWithPrefix(config),
+                        ImmutableList.of(currentUser.getShortUserName())),
+                () ->
+                    getServiceAccountToImpersonateForUserGroup(
+                        GROUP_IMPERSONATION_SERVICE_ACCOUNT_SUFFIX
+                            .withPrefixes(CONFIG_KEY_PREFIXES)
+                            .getPropsWithPrefix(config),
+                        ImmutableList.copyOf(currentUser.getGroupNames())),
+                (Supplier<Optional<String>>)
+                    () ->
+                        Optional.ofNullable(
+                            IMPERSONATION_SERVICE_ACCOUNT_SUFFIX
+                                .withPrefixes(CONFIG_KEY_PREFIXES)
+                                .get(config, config::get)))
+            .map(Supplier::get)
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .filter(sa -> !isNullOrEmpty(sa))
+            .findFirst();
```

Comment:
Not super familiar with lambda in java, but will this always return some form of Optional object even if any filter above returns nothing?

Reply:
Yes, it will return `Optinal.empty()` if nothing is found.